Add the instructions on how to configure the operator CRs (#2273)
knative/docs: docs/install/operator/configuring-serving-cr.md

---
title: "Configuring the Serving Operator Custom Resource"
weight: 20
type: "docs"
aliases:
- /docs/operator/configuring-serving-cr/
---

The Knative Serving operator can be configured with these options:

> - [All the ConfigMaps](#all-the-configmaps)

Reviewer: I just read this, and it reads strangely:

"The Knative Serving operator can be configured with the option 'all the ConfigMaps'."

Maybe "Service Configuration by ConfigMap"?

You probably also want to change the last one to 'System Resource Settings', rather than saying "The Knative Serving operator can be configured with the option 'managing resources for containers'".

```

### Download images individually without secrets:

If your custom image links are not defined in a uniform format by default, you will need to individually include each
link in the CR.

For example, to given the following images:

| Container | Docker Image |
|----|----|
| `activator` | `docker.io/knative-images-repo1/activator:v0.13.0` |
| `autoscaler` | `docker.io/knative-images-repo2/autoscaler:v0.13.0` |
| `controller` | `docker.io/knative-images-repo3/controller:v0.13.0` |
| `webhook` | `docker.io/knative-images-repo4/webhook:v0.13.0` |
| `autoscaler-hpa` | `docker.io/knative-images-repo5/autoscaler-hpa:v0.13.0` |
| `networking-istio` | `docker.io/knative-images-repo6/prefix-networking-istio:v0.13.0` |
| `queue-proxy` | `docker.io/knative-images-repo7/queue-proxy-suffix:v0.13.0` |

The operator CR should be modified to include the full list:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  registry:
    override:
      activator: docker.io/knative-images-repo1/activator:v0.13.0
      autoscaler: docker.io/knative-images-repo2/autoscaler:v0.13.0

>         controller: docker.io/knative-images-repo3/controller:v0.13.0

Reviewer: ```suggestion
      controller: docker.io/knative-images-repo3/controller:v0.13.0
```

spec:
  registry:
    ...
    imagePullSecrets:
      - name: regcred
      - name: regcred-2
      ...
```

## SSL certificate for controller

To [enable tag to digest resolution](https://knative.dev/development/serving/tag-resolution/), the Knative Serving controller needs to access the container registry.
To allow the controller to trust a self-signed registry cert, you can use the Operator to specify the certificate using a ConfigMap or Secret.

Specify the following fields in `spec.controller-custom-certs` to select a custom registry certificate:

- `name`: the name of the ConfigMap or Secret.
- `type`: either the string "ConfigMap" or "Secret".

If you create a ConfigMap named `testCert` containing the certificate, change your CR:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  controller-custom-certs:
    name: testCert
    type: ConfigMap
```

## Configuration of Knative ingress gateway

To set up custom ingress gateway, follow [**Step 1: Create Gateway Service and Deployment Instance**](https://knative.dev/development/serving/setting-up-custom-ingress-gateway/).

> **Step 2: Update the Knative gateway**

Reviewer: ```suggestion
### Step 2: Update the Knative gateway
```

Update `spec.knative-ingress-gateway` to select the labels of the new ingress gateway:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  knative-ingress-gateway:
    selector:
      custom: ingressgateway
```

> **Step 3: Update Gateway ConfigMap**

Reviewer: ```suggestion
### Step 3: Update Gateway ConfigMap
```

The following configuration specifies a replica count of 3 for the controllers and a minimum of 3 activators (which may scale higher if needed):

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  high-availability:
    replicas: 3
```

## Managing resources for containers

The operator custom resource allows you allows you to configure resource requests and limits for the Knative system containers.
Requests and limits may be configured for the following containers: `activator`, `autoscaler`, `controller`, `webhook`, `autoscaler-hpa`,
`networking-istio` and `queue-proxy`.

To override resource settings for a specific container, create an entry in the `spec.resources` list with the container name and the [Kubernetes resource settings](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container).

For example, the following KnativeServing resource configures the `activator` to request 0.3 CPU and 100MB of RAM, and sets hard limits of 1 CPU, 250MB RAM, and 4GB of local storage:

> \```

Reviewer: You need to remove the "\" here and on line 359. I can't use suggestion blocks for that, because they use the <code>```</code> shortcode.

Author: done

apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  resources:
  - container: activator
    requests:
      cpu: 300m
      memory: 100Mi
    limits:
      cpu: 1000m
      memory: 250Mi
      ephemeral-storage: 4Gi
\```

If you would like to add another container `autoscaler` with the same configuration, you need to change your CR as below:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  resources:

>   - container: activator
>       requests:
>         cpu: 30m
>         memory: 40Mi
>         storage: 1Gi
>         ephemeral-storage: 2Gi
>       limits:
>         cpu: 300m
>         memory: 400Mi
>         storage: 2Gi
>         ephemeral-storage: 4Gi
>   - container: autoscaler
>       requests:
>         cpu: 30m
>         memory: 40Mi
>         storage: 1Gi
>         ephemeral-storage: 2Gi
>       limits:
>         cpu: 300m
>         memory: 400Mi
>         storage: 2Gi
>         ephemeral-storage: 4Gi

Reviewer: The YAML indentation here is off; I also made the quantities in this example align with the one above.

```suggestion
  - container: activator
    requests:
      cpu: 300m
      memory: 100Mi
    limits:
      cpu: 1000m
      memory: 250Mi
      ephemeral-storage: 4Gi
  - container: autoscaler
    requests:
      cpu: 300m
      memory: 100Mi
    limits:
      cpu: 1000m
      memory: 250Mi
      ephemeral-storage: 4Gi
```

The `KnativeServing` custom resource allows you to set values for these ConfigMaps via the operator. Knative Serving has multiple ConfigMaps named with the prefix
`config-`. The `spec.config` in `KnativeServing` has one entry `<name>` for each ConfigMap named `config-<name>`, with a value which will be used for the ConfigMap's `data`.

In the [setup a custom domain example](https://knative.dev/development/serving/using-a-custom-domain/), you can see the content of the ConfigMap
`config-domain` is:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-domain
  namespace: knative-serving
data:
  example.org: |
    selector:
      app: prod
  example.com: ""
```

Using the operator, specify the ConfigMap `config-domain` using the operator CR:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:
    domain:
      example.org: |
        selector:
          app: prod
      example.com: ""
```

Next, save the CR into a file named `operator-cr.yaml`, and run the command:

```
kubectl apply -f operator-cr.yaml
```

> You can apply values to multiple ConfigMaps. This exmaple sets `stable-window` to 60s in `config-autoscaler` as well as specifying `config-domain`:

Reviewer 2: <!--typo-bot-->
Found potential misspelling(s):
```suggestion
You can apply values to multiple ConfigMaps. This example sets `stable-window` to 60s in `config-autoscaler` as well as specifying `config-domain`:
```

> You can apply values to multiple ConfigMaps. This exmaple sets `stable-window` to 60s in `config-autoscaler` as well as specifying `config-domain`:

Reviewer 2: <!--typo-bot-->
Found potential misspelling(s):
```suggestion
You can apply values to multiple ConfigMaps. This example sets `stable-window` to 60s in `config-autoscaler` as well as specifying `config-domain`:
```

> You can apply values to multiple ConfigMaps. This exmaple sets `stable-window` to 60s in `config-autoscaler` as well as specifying `config-domain`:

Reviewer 2: <!--typo-bot-->
Found potential misspelling(s):
```suggestion
You can apply values to multiple ConfigMaps. This example sets `stable-window` to 60s in `config-autoscaler` as well as specifying `config-domain`:
```

The `KnativeServing` custom resource allows you to set values for these ConfigMaps via the operator. Knative Serving has multiple ConfigMaps named with the prefix
`config-`. The `spec.config` in `KnativeServing` has one entry `<name>` for each ConfigMap named `config-<name>`, with a value which will be used for the ConfigMap's `data`.

In the [setup a custom domain example](https://knative.dev/development/serving/using-a-custom-domain/), you can see the content of the ConfigMap
`config-domain` is:

```
apiVersion: v1
kind: ConfigMap
metadata:
  name: config-domain
  namespace: knative-serving
data:
  example.org: |
    selector:
      app: prod
  example.com: ""
```

Using the operator, specify the ConfigMap `config-domain` using the operator CR:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:
    domain:
      example.org: |
        selector:
          app: prod
      example.com: ""
```

Next, save the CR into a file named `operator-cr.yaml`, and run the command:

```
kubectl apply -f operator-cr.yaml
```

> You can apply values to multiple ConfigMaps. This exmaple sets `stable-window` to 60s in `config-autoscaler` as well as specifying `config-domain`:

Reviewer 2: <!--typo-bot-->
Found potential misspelling(s):
```suggestion
You can apply values to multiple ConfigMaps. This example sets `stable-window` to 60s in `config-autoscaler` as well as specifying `config-domain`:
```

    selector:
      app: prod
  example.com: ""
```

To specify the ConfigMap `config-domain`, you can change the content of the operator CR into:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:
    domain:
      example.org: |
        selector:
          app: prod
      example.com: ""
```

Next, save the CR into a file named `operator-cr.yaml`, and run the command:

```
kubectl apply -f operator-cr.yaml
```

If you want to change another ConfigMap, e.g. `config-autoscaler`, by specifying `stable-window` to `60s`. Continue to
change your operator CR into:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  config:
    domain:
      example.org: |
        selector:
          app: prod
      example.com: ""
    autoscaler:
      stable-window: "60s"
```

Then, save the content in a file named `operator-cr.yaml`, and run the command:

```
kubectl apply -f operator-cr.yaml
```

All the ConfigMaps should be defined under the same namespace as the operator CR. You can use the operator CR as the

> unique entry point to edit all of them.

Author: It should be `All the ConfigMaps are created in the same namespace as the operator CR`

> unique entry point to edit all of them.

Reviewer: Sorry, my question was (assuming that you know nothing about how Knative is built/configured), what should the the operator do with the information about where ConfigMaps might exist?

```

The key in `spec.config.istio` is in the format of `gateway.{{gateway_namespace}}.{{gateway_name}}`.

## Configuration of cluster local gateway:

We use the field `cluster-local-gateway` to override the the gateway cluster-local-gateway. We only support the field
`selector` to define the selector for the local gateway.

**Default local gateway name**:

Go through the guide [here](https://knative.dev/development/install/installing-istio/#updating-your-install-to-use-cluster-local-gateway) to use local cluster gateway.

After following the above step, your service and deployment for the local gateway are both named `cluster-local-gateway`.
You only need to configure the operator CR as below:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  registry:
    cluster-local-gateway:
      selector:
        istio: cluster-local-gateway
```

You can even skip the above change, since there is a gateway called `cluster-local-gateway`, which has
`istio: cluster-local-gateway` as the default selector. If the operator CR does not define the section

> cluster-local-gateway, the default `istio: cluster-local-gateway` of the gateway cluster-local-gateway will be chosen.

Author: I will remove this operation CR change for this section.

cluster-local-gateway.

If you name both of the service and deployment after `custom-local-gateway` in the namespace `istio-system`, with the
label `custom: custom-local-gateway`, the operator CR should be like:

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  registry:
    cluster-local-gateway:
      selector:
        custom: custom-local-gateway
    config:
      istio:
        local-gateway.knative-serving.cluster-local-gateway: "custom-local-gateway.istio-system.svc.cluster.local"
```

## High availability:

Knative Serving Operator CR provides the capability to configure the high availability for the following `Deployment`
resources: `controller`, `autoscaler-hpa` & `networking-istio`, and all the `HorizontalPodAutoscaler` resources. A field
named `high-availability` is defined under the section `spec` to configure the high availability control plane. In the
latest release of Knative Serving Operator, you can specify the minimum number of replicas for the `Deployment` and
`HorizontalPodAutoscaler` resources.

> If you want to specify 3 as the minimum number of replicas for three of the `Deployment` resources and the `HorizontalPodAutoscaler`
> resources, change your CR into:

Reviewer: ```suggestion
The following configuration specifies a replica count of 3 for the controllers and a minimum of 3 activators (which may scale higher if needed):
```

```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  high-availability:
    replicas: 3
```

## Managing resources for containers

> Knative Serving Operator CR allows you to configure the resources for the containers defined in the `Deployment` resource.
> Each `Deployment` resource has only one container with the same deployment name. The following containers can be configured
> in terms of the required and maximum resources to be allocated: `activator`, `autoscaler`, `controller`, `webhook`, `autoscaler-hpa`,
> `networking-istio` and `queue-proxy`. The field called `resources` holds the stanza of a list of containers for resource
> configuration. The field `require` defines how much resource the container requires. The field `limit` defines the maximum
> resource the container can use. Within each field, each container can be configured individually with the following
> parameters: `memory`, `cpu`, `storage` and `ephemeral-storage`. The value in `require` should always be smaller than the
> corresponding value in `limit`. The field `container` is used to specify the container name.

Reviewer: ```suggestion
The operator custom resource allows you allows you to configure resource requests and limits for the Knative system containers.
Requests and limits may be configured for the following containers: `activator`, `autoscaler`, `controller`, `webhook`, `autoscaler-hpa`,
`networking-istio` and `queue-proxy`.

To override resource settings for a specific container, create an entry in the `spec.resources` list with the container name and the [Kubernetes resource settings](https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#resource-requests-and-limits-of-pod-and-container).
```

> Suppose you would like to configure the container called `activator` with the required resources:
> 
> ```
> cpu: 30m
> memory: 40Mi
> storage: 1Gi
> ephemeral-storage: 2Gi
> ```
> 
> , and the limit resources:
> 
> ```
> cpu: 300m
> memory: 4000Mi
> storage: 2Gi
> ephemeral-storage: 4Gi
> ```
> 
> Here is how you should write your CR:
> 
> ```
> apiVersion: operator.knative.dev/v1alpha1
> kind: KnativeServing
> metadata:
>   name: knative-serving
>   namespace: knative-serving
> spec:
>   resources:
>   - container: activator
>       requests:
>         cpu: 30m
>         memory: 40Mi
>         storage: 1Gi
>         ephemeral-storage: 2Gi
>       limits:
>         cpu: 300m
>         memory: 400Mi
>         storage: 2Gi
>         ephemeral-storage: 4Gi
> ```

Reviewer: ```suggestion
For example, the following KnativeServing resource configures the `activator` to request 0.3 CPU and 100MB of RAM, and sets hard limits of 1 CPU, 250MB RAM, and 4GB of local storage:

\```
apiVersion: operator.knative.dev/v1alpha1
kind: KnativeServing
metadata:
  name: knative-serving
  namespace: knative-serving
spec:
  resources:
  - container: activator
    requests:
      cpu: 300m
      memory: 100Mi
    limits:
      cpu: 1000m
      memory: 250Mi
      ephemeral-storage: 4Gi
\```
```
